tests/compatibility_test.sh

```diff
@@ -0,0 +1,324 @@
+    expect_log " <-- target platform (${platform}) ${error_string}"
+    expect_log 'FAILED: Build did NOT complete successfully'
+  done
+}
+
+# Validates that we can express targets being compatible with A _or_ B.
+function test_any_of_logic() {
+  cat >> target_skipping/BUILD <<EOF
+sh_test(
+    name = "pass_on_foo1_or_foo2_but_not_on_foo3",
+    srcs = [":pass.sh"],
+    target_compatible_with = compatibility.any_of(":foo1", ":foo2"),
+)
+EOF
+
+  ensure_that_target_builds_for_platforms \
+    //target_skipping:pass_on_foo1_or_foo2_but_not_on_foo3 \
+    //target_skipping:foo1_bar1_platform \
+    //target_skipping:foo2_bar1_platform \
+    //target_skipping:foo2_bar2_platform
+
+  ensure_that_target_doesnt_build_for_platforms \
+    //target_skipping:pass_on_foo1_or_foo2_but_not_on_foo3 \
+    "didn't satisfy constraint //lib/compatibility:any_of$" \
+    //target_skipping:foo3_platform \
+    //target_skipping:bar1_platform
+}
+
+# Validates that we can express targets being compatible with everything _but_
+# A and B.
+function test_none_of_logic() {
+  cat >> target_skipping/BUILD <<EOF
+sh_test(
+    name = "pass_on_everything_but_foo1_and_foo2",
+    srcs = [":pass.sh"],
+    target_compatible_with = compatibility.none_of(":foo1", ":foo2"),
+)
+EOF
+
+  ensure_that_target_builds_for_platforms \
+    //target_skipping:pass_on_everything_but_foo1_and_foo2 \
+    //target_skipping:foo3_platform \
+    //target_skipping:bar1_platform
+
+  ensure_that_target_doesnt_build_for_platforms \
+    //target_skipping:pass_on_everything_but_foo1_and_foo2 \
+    "didn't satisfy constraint //lib/compatibility:none_of$" \
+    //target_skipping:foo1_bar1_platform \
+    //target_skipping:foo2_bar1_platform \
+    //target_skipping:foo2_bar2_platform
+}
+
+# Validates that we can express targets being compatible with _only_ A and B,
+# and nothing else.
+function test_all_of_logic() {
+  cat >> target_skipping/BUILD <<EOF
+sh_test(
+    name = "pass_on_only_foo1_and_bar1",
+    srcs = [":pass.sh"],
+    target_compatible_with = compatibility.all_of(":foo1", ":bar1"),
+)
+EOF
+
+  ensure_that_target_builds_for_platforms \
+    //target_skipping:pass_on_only_foo1_and_bar1 \
+    //target_skipping:foo1_bar1_platform
+
+  ensure_that_target_doesnt_build_for_platforms \
+    //target_skipping:pass_on_only_foo1_and_bar1 \
+    "didn't satisfy constraints\\? \\[\\?//lib/compatibility:all_of_" \
+    //target_skipping:foo2_bar1_platform \
+    //target_skipping:foo2_bar2_platform \
+    //target_skipping:foo3_platform \
+    //target_skipping:bar1_platform
+}
+
+# Validates that we can express composed incompatibility.
+function test_composition() {
+  cat >> target_skipping/BUILD <<EOF
+sh_test(
+    name = "pass_on_foo1_or_foo2_but_not_bar1",
+    srcs = [":pass.sh"],
+    target_compatible_with = compatibility.any_of(
+        ":foo1",
+        ":foo2",
+    ) + compatibility.none_of(
+        ":bar1",
+    ),
```

Comment:
So `any_of(foo1)+none_of(foo1)` is always false, right?
What about `all+none`, `any+all`, `any+any`, `all+all`, `none+none`, they all work right?

Reply:
> So any_of(foo1)+none_of(foo1) is always false, right?

I have not added a test to explicitly test this, but I don't see any way that it would ever evaluate to true.

> What about all+none, any+all, any+any, all+all, none+none, they all work right?

Theoretically yes. The challenge at the moment is that because of bazel's dislike of duplicated labels in label lists, we can't do, say, any+any, all+all, or none+none. I think the only way to make that work is to either:
1. Figure out how to make bazel not complain about duplicate labels in a label list, or
2. Accept a `name`-like parameter for these helpers that will create new `constraint_value` targets for that invocation.

Reply:
I've got a feeling there is something fishy going on, and I'm not able to put my finger on it. There is negation without universal set and `+` seems to mean the same as conjunction. The feeling is you're not using the right set of primitives. But I'm failing to figure out what those should be.

For example using `+` for conjunction, `not` for single negation and `any` for disjunction. This way you should be able to write any boolean expression as something like `= any(a,b,not(c)) + any(c,d,not(e))`. But then `a+b` only works for constraints. Should you also have `as_constraints`, that wraps a config value.

I can't really tell what's more usable. So far I've seen the most places need just conjunction of constraints and some more rare places a disjunction of constraints.

Reply:
Thinking out loud: maybe opposite approach is the solution: a single function `match([a,b],c,[d,e], noneof=[f,g])` which says `(a and b) and c and (d or e) and (not f) and (not g)`.

With current situation about composition - I would be more happy to make no guarantees about it, because they seem to be quite "flaky". Maybe even better, breaking composition on purpose, by using that single `incompatible` constraint value.

Reply:
Filed bazelbuild/bazel#16274 to see what the usual suspects think about it.

Reply:
One more thought. What if we allow config_values in the list? I think this is doable in Bazel. @katre?

This could simplify a lot what's going on here, specially `all_of` would become unnecessary. I believe the initial design decision was not to mix constraints and configuration, but we could revisit that.

Reply:
One of the pieces of wisdom that @gregestren imparted on me early on is that if something can reasonably simply be implemented in Starlark instead of native bazel, then Starlark is the preferred option. I.e. keep things out of bazel if there's a reasonable approach to doing it in Starlark. I have not yet found anything that has convinced me otherwise. In this case, I'm not convinced that we have to make bazel's implementation more complicated just to make the Starlark implementation here simpler.

Of course I will let you all be the judge of whether or not using a series of `select()` statements is a reasonable approach to avoid making bazel more complicated.

Reply:
What is meant by `config_value`? Is `config_setting` meant?

Reply:
I'm assuming so. I myself can't for the life of me keep them straight. The fact that `select()` accepts config_**setting**s and constraint_**value**s makes me accidentally mix the two very frequently.

Reply:
Using `config_setting` in `target_compatible_with` is confusing: what does it mean if a target is `target_compatible_with` a setting that checks whether flag `--compilation_mode` is set to `OPT`? Either we define semantics for the full set of `config_setting` behaviors, or we're just using it to group `constraint_values`, and I'd argue that isn't worthwhile.

I agree that the similarity and differences between `config_setting` and `constraint_value` is unfortunate, but even several years later I haven't been able to come up with better names.

Reply:
I agree that with bare bazel, there aren't great examples for `target_compatible_with` and `config_setting`.

I suspect most of the interesting use cases come from user-defined `config_setting` targets. For example, we internally use `target_compatible_with` to mark things as only compatible with `clang` vs. `gcc`. That doesn't actually have anything to do with target platforms, but it's a good mechanism. We also use this to mark code as compatible with certain versions of Python. We don't encode the Python version in the target platform since everything is hermetic. In short, we use it to make assertions about code compatibility with toolchains. The only way that is possible today is with `target_compatible_with` and `config_setting`.

Regardless, I do technically like your `match()` idea. I just need to figure out how to make it more understandable when you don't know what the function does. That's really the only reason I haven't switched to it yet.

In the meantime, I've switched the implementation here to use `@platforms//:incompatible` for everything. The tests all pass when run against bazel HEAD.

I will keep thinking about how to make the `match()` idea work well.

Reply:
@comius , I spent some time implementing your `match` idea, but I'm not convinced that it'll be ergonomic. Starting with this:
```
target_compatible_with = compatibility.match([":a", ":b"]),  # This means a _or_ b.
target_compatible_with = compatibility.match(":a", ":b"),  # This means a _and_ b.
```
makes me think the difference is too subtle. Making it more verbose:
```
target_compatible_with = compatibility.match_all(  # This means a _or_ b.
    compatibility.or_group(":a", ":b"),
),
target_compatible_with = compatibility.match_all(":a", ":b"),  # This means a _and_ b.
```
makes it easier to understand But then we're not really saving the user much over doing:
```
target_compatible_with = compatibility.any_of(":a", ":b"),  # This means a _or_ b.
target_compatible_with = compatibility.all_of(":a", ":b"),  # This means a _and_ b.
```

With bazel HEAD, it's actually pretty easy now to compose the `match([a,b],c,[d,e], noneof=[f,g])` example from earlier.
```
target_compatible_with = compatibility.any_of(a, b) + \
                         compatibility.all_of(c) + \
                         compatibility.any_of(d, e) + \
                         compatibility.none_of(f, g),
```

WDYT about the current state of things?

Reply:
I also think different behavior for list of conditions vs. a single set is far too subtle. People will get that wrong.
any_of and all_of are excellent names.

Reply:
@comius , any thoughts?

I can disable the tests for pre-6.0 versions of bazel. But that would mean not running tests at all for now.

Reply:
@comius , WDYT?

Reply:
@comius , ping